# 🧪 Lab — Can You Trust This Dataset?

## Overview

You have been given a dataset containing e-commerce sales data.

Your manager has already used this dataset to report revenue metrics to leadership.

Your job is to determine:

> **Can this dataset be trusted for analysis?**

---

## IMPORTANT

To get the dataset, navigate to the `./data/raw/` folder. 

You should see a Python file called: `generate_sales_data.py`.

**IN YOUR TERMINAL** inside the `./data/raw/` folder run the following command:

```
python3 generate_sales_data.py
```

You should then have a `sales.csv` file in the folder after it runs.

In [5]:
# Run your imports here`
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
import json
import os

os.chdir(r"C:\Users\evovo\lab-09-data-validation")

exec(open('data/raw/generate_sales_data.py').read())
sales = generate_sales_data()
sales.to_csv('data/raw/sales_data.csv', index=False)
sales = pd.read_csv('data/raw/sales_data.csv')

print("Data loaded successfully!")
print("Shape:", sales.shape)
display(sales.head())

print("All libraries loaded successfully!")

sales.csv generated successfully.
Data loaded successfully!
Shape: (305, 10)


<string>:156: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '['97.73' '218.37' '-131.93' '262.68' '183.89' '399.22' '132.71' '347.35'
 '160.21' '175.55']' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
<string>:156: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '['97.73' '218.37' '-131.93' '262.68' '183.89' '399.22' '132.71' '347.35'
 '160.21' '175.55']' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.


,order_id,customer_id,product_id,revenue,quantity,unit_price,status,payment_method,region,order_date
0,1000,52.0,217,124.55,3,52.18,Complete,Unknown,East,2023-02-26
1,1001,15.0,224,108.94,0,75.70,Cancelled,Debit Card,west,2023-03-07
2,1002,72.0,209,286.86,1,65.15,Returned,Credit Card,West,2023-11-26
3,1003,61.0,221,259.37,3,47.06,Complete,Credit Card,East,2023-04-08
4,1004,21.0,225,246.81,1,33.59,Returned,Gift Card,West,2023-05-24


All libraries loaded successfully!


Load the dataset.

Name the dataframe `sales`

In [6]:
sales = pd.read_csv('data/raw/sales_data.csv')
display(sales.head())


,order_id,customer_id,product_id,revenue,quantity,unit_price,status,payment_method,region,order_date
0,1000,52.0,217,124.55,3,52.18,Complete,Unknown,East,2023-02-26
1,1001,15.0,224,108.94,0,75.70,Cancelled,Debit Card,west,2023-03-07
2,1002,72.0,209,286.86,1,65.15,Returned,Credit Card,West,2023-11-26
3,1003,61.0,221,259.37,3,47.06,Complete,Credit Card,East,2023-04-08
4,1004,21.0,225,246.81,1,33.59,Returned,Gift Card,West,2023-05-24


---

### Initial Analysis

In [7]:
# TODO: Calculate total revenue
total_revenue = (sales["quantity"] * sales["unit_price"]).sum()
print(f"Total revenue: ${total_revenue:,.2f}")
# TODO: Calculate average order value
avg_order_value = (sales["quantity"] * sales["unit_price"]).mean()
print(f"Average order value: ${avg_order_value:,.2f}")
# TODO: Calculate revenue by status
revenue_by_status = (
    sales.groupby("status")
    .apply(lambda g: (g["quantity"] * g["unit_price"]).sum(), include_groups=False)
    .rename("total_revenue")
    .reset_index()
)
print("\nRevenue by status:")
display(revenue_by_status)

Total revenue: $42,830.94
Average order value: $140.43

Revenue by status:


,status,total_revenue
0,COMPLETE,319.15
1,CANCELLED,605.83
2,Cancelled,6790.45
3,Cancelled,288.27
4,Complete,29360.37
5,Returned,3087.93
6,Returned,1160.86
7,complete,309.87
8,completed,38.50
9,return,109.14


---

### Data Inspection

Call the usual Pandas methods to inspect a dataset.

In [8]:
display(sales.head())           
display(sales.tail())          
print(sales.shape)              
sales.info()                    
display(sales.describe())       
print(sales.columns.tolist())   
print(sales.isnull().sum())     
print(sales.dtypes)  

,order_id,customer_id,product_id,revenue,quantity,unit_price,status,payment_method,region,order_date
0,1000,52.0,217,124.55,3,52.18,Complete,Unknown,East,2023-02-26
1,1001,15.0,224,108.94,0,75.70,Cancelled,Debit Card,west,2023-03-07
2,1002,72.0,209,286.86,1,65.15,Returned,Credit Card,West,2023-11-26
3,1003,61.0,221,259.37,3,47.06,Complete,Credit Card,East,2023-04-08
4,1004,21.0,225,246.81,1,33.59,Returned,Gift Card,West,2023-05-24


,order_id,customer_id,product_id,revenue,quantity,unit_price,status,payment_method,region,order_date
300,1038,73.0,224,191.84,1,31.26,Complete,Unknown,Midwest,2023-09-11
301,1161,44.0,203,118.13,5,44.25,Complete,Gift Card,Midwest,2023-05-08
302,1288,30.0,211,305.19,4,39.01,Returned,credit card,East,2023-05-12
303,1136,66.0,211,347.35,4,57.71,Complete,Gift Card,East,2023-11-10
304,1095,41.0,9999,205.06,1,57.38,Complete,Debit Card,South,2023-06-11


(305, 10)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 305 entries, 0 to 304
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_id         305 non-null    int64  
 1   customer_id      293 non-null    float64
 2   product_id       305 non-null    int64  
 3   revenue          305 non-null    object 
 4   quantity         305 non-null    int64  
 5   unit_price       305 non-null    float64
 6   status           305 non-null    object 
 7   payment_method   305 non-null    object 
 8    region          305 non-null    object 
 9   order_date       305 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 24.0+ KB


,order_id,customer_id,product_id,quantity,unit_price
count,305.000000,293.000000,305.000000,305.000000,305.000000
mean,1149.363934,38.709898,343.629508,2.819672,49.615607
std,86.692792,23.068268,1114.916428,1.458958,20.288912
min,1000.000000,-1.000000,200.000000,0.000000,-96.570000
25%,1075.000000,20.000000,208.000000,2.000000,40.240000
50%,1149.000000,38.000000,216.000000,3.000000,49.600000
75%,1224.000000,60.000000,223.000000,4.000000,60.170000
max,1299.000000,79.000000,9999.000000,5.000000,97.510000


['order_id', 'customer_id', 'product_id', 'revenue', 'quantity', 'unit_price', 'status', 'payment_method ', ' region', 'order_date']
order_id            0
customer_id        12
product_id          0
revenue             0
quantity            0
unit_price          0
status              0
payment_method      0
 region             0
order_date          0
dtype: int64
order_id             int64
customer_id        float64
product_id           int64
revenue             object
quantity             int64
unit_price         float64
status              object
payment_method      object
 region             object
order_date          object
dtype: object


---

### Idenfity Issues

Feel free to break this up across multiple cells

In [9]:
# TODO: Check for missing values
print("=== Missing Values ===")
print(sales.isnull().sum())

# TODO: Check for duplicates
print("\n=== Duplicates ===")
print("Duplicate rows:", sales.duplicated().sum())
print("Duplicate order IDs:", sales["order_id"].duplicated().sum())
# TODO: Check for invalid revenue
print("\n=== Invalid Revenue ===")
sales["revenue_clean"] = pd.to_numeric(
    sales["revenue"].astype(str).str.replace(r"[$,]", "", regex=True), 
    errors="coerce")

print("Negative revenue:", (sales["revenue_clean"] < 0).sum())
print("Zero revenue:", (sales["revenue_clean"] == 0).sum())
print("Null revenue after cleaning:", sales["revenue_clean"].isnull().sum())

# TODO: Check for invalid quantities
print("\n=== Invalid Quantities ===")
print("Zero quantity:", (sales["quantity"] == 0).sum())
print("Negative quantity:", (sales["quantity"] < 0).sum())
# TODO: Check for inconsistent categories
print("\n=== Inconsistent Categories ===")
print("Unique status values:")
print(sales["status"].value_counts())
print("\nUnique payment methods:")
print(sales["payment_method "].value_counts())
# TODO: Check for invalid dates
print("\n=== Invalid Dates ===")
sales["order_date"] = pd.to_datetime(sales["order_date"], errors="coerce")
print("Future dates:", (sales["order_date"] > pd.Timestamp.today()).sum())
print("Null dates:", sales["order_date"].isnull().sum())

=== Missing Values ===
order_id            0
customer_id        12
product_id          0
revenue             0
quantity            0
unit_price          0
status              0
payment_method      0
 region             0
order_date          0
dtype: int64

=== Duplicates ===
Duplicate rows: 4
Duplicate order IDs: 17

=== Invalid Revenue ===
Negative revenue: 8
Zero revenue: 0
Null revenue after cleaning: 0

=== Invalid Quantities ===
Zero quantity: 6
Negative quantity: 0

=== Inconsistent Categories ===
Unique status values:
status
Complete      209
Cancelled      48
Returned       23
Returned        6
CANCELLED       4
unknown         3
Cancelled       3
 COMPLETE       3
complete        3
return          2
completed       1
Name: count, dtype: int64

Unique payment methods:
payment_method 
Gift Card        86
Credit Card      73
Debit Card       66
PayPal           58
paypal            5
Unknown           4
giftcard          4
credit card       3
CreditCard        3
Pay Pal          

---

### Create Validation Columns

In [10]:
# TODO: Create validation flags

# Example:
# sales["missing_customer_id"] = ...
sales["missing_customer_id"] = sales["customer_id"].isnull()
sales["negative_revenue"] = sales["revenue_clean"] < 0
sales["zero_quantity"] = sales["quantity"] == 0
sales["future_date"] = sales["order_date"] > pd.Timestamp.today()
sales["duplicate_order_id"] = sales["order_id"].duplicated(keep=False)
sales["invalid_unit_price"] = sales["unit_price"] <= 0
sales["null_revenue"] = sales["revenue_clean"].isnull()
sales.head()

,order_id,customer_id,product_id,revenue,quantity,unit_price,status,payment_method,region,order_date,revenue_clean,missing_customer_id,negative_revenue,zero_quantity,future_date,duplicate_order_id,invalid_unit_price,null_revenue
0,1000,52.0,217,124.55,3,52.18,Complete,Unknown,East,2023-02-26,124.55,False,False,False,False,False,False,False
1,1001,15.0,224,108.94,0,75.70,Cancelled,Debit Card,west,2023-03-07,108.94,False,False,True,False,False,False,False
2,1002,72.0,209,286.86,1,65.15,Returned,Credit Card,West,2023-11-26,286.86,False,False,False,False,False,False,False
3,1003,61.0,221,259.37,3,47.06,Complete,Credit Card,East,2023-04-08,259.37,False,False,False,False,False,False,False
4,1004,21.0,225,246.81,1,33.59,Returned,Gift Card,West,2023-05-24,246.81,False,False,False,False,False,False,False


---

### Summarize Issues



In [11]:
# TODO: Summarize validation columns
flag_cols = [
    "missing_customer_id",
    "negative_revenue",
    "zero_quantity",
    "future_date",
    "duplicate_order_id",
    "invalid_unit_price",
    "null_revenue"
]
print("=== Validation Summary ===")
summary = pd.DataFrame({
    "flag": flag_cols,
    "issue_count": [sales[flag].sum() for flag in flag_cols],
    "percentage": [f"{(sales[flag].sum() / len(sales) * 100):.1f}%" for flag in flag_cols]
})
display(summary)

=== Validation Summary ===


,flag,issue_count,percentage
0,missing_customer_id,12,3.9%
1,negative_revenue,8,2.6%
2,zero_quantity,6,2.0%
3,future_date,6,2.0%
4,duplicate_order_id,33,10.8%
5,invalid_unit_price,6,2.0%
6,null_revenue,0,0.0%


---

### Assertions

In [12]:
# TODO: Write assertions
print("=== Running Assertions ===")

try:
    assert sales["order_id"].notnull().all(), "order_id has null values"
    print("order_id has no nulls")
except AssertionError as e:
    print(f"FAIL: {e}")

try:
    assert (sales["quantity"] > 0).all(), "quantity has zero or negative values"
    print("quantity all positive")
except AssertionError as e:
    print(f"FAIL: {e}")
try:
    assert (sales["unit_price"] > 0).all(), "unit_price has zero or negative values"
    print("unit_price all positive")
except AssertionError as e:
    print(f"FAIL: {e}")

try:
    assert sales["revenue_clean"].notnull().all(), "revenue has null or unparseable values"
    print("revenue has no nulls")
except AssertionError as e:
    print(f"FAIL: {e}")

try:
    assert (sales["revenue_clean"] >= 0).all(), "revenue has negative values"
    print("revenue all non-negative")
except AssertionError as e:
    print(f"FAIL: {e}")

try:
    assert (sales["order_date"] <= pd.Timestamp.today()).all(), "order_date has future dates"
    print("order_date no future dates")
except AssertionError as e:
    print(f"FAIL: {e}")

try:
    assert not sales.duplicated().any(), "dataset has duplicate rows"
    print("no duplicate rows")
except AssertionError as e:
    print(f"FAIL: {e}")

print("\n=== Assertions Complete ===")


=== Running Assertions ===
order_id has no nulls
FAIL: quantity has zero or negative values
FAIL: unit_price has zero or negative values
revenue has no nulls
FAIL: revenue has negative values
FAIL: order_date has future dates
FAIL: dataset has duplicate rows

=== Assertions Complete ===


---

### Now clean your dataset

Save it as `clean_sales`

In [13]:
# TODO: Filter out problematic rows
sales["has_any_issue"] = sales[flag_cols].any(axis=1)
clean_sales = sales[~sales["has_any_issue"]].copy()
clean_sales = clean_sales.reset_index(drop=True)


In [14]:
clean_sales.head()

,order_id,customer_id,product_id,revenue,quantity,unit_price,status,payment_method,region,order_date,revenue_clean,missing_customer_id,negative_revenue,zero_quantity,future_date,duplicate_order_id,invalid_unit_price,null_revenue,has_any_issue
0,1000,52.0,217,124.55,3,52.18,Complete,Unknown,East,2023-02-26,124.55,False,False,False,False,False,False,False,False
1,1002,72.0,209,286.86,1,65.15,Returned,Credit Card,West,2023-11-26,286.86,False,False,False,False,False,False,False,False
2,1003,61.0,221,259.37,3,47.06,Complete,Credit Card,East,2023-04-08,259.37,False,False,False,False,False,False,False,False
3,1004,21.0,225,246.81,1,33.59,Returned,Gift Card,West,2023-05-24,246.81,False,False,False,False,False,False,False,False
4,1005,75.0,202,6178.25,5,14.43,Cancelled,Gift Card,West,2023-05-05,6178.25,False,False,False,False,False,False,False,False


---

### Re-run the analysis

Reference the lab `README` to remember which metrics to investigate.

In [15]:
# TODO: Recalculate metrics
print("=== BEFORE CLEANING (original sales) ===")
total_revenue_before = (sales["quantity"] * sales["unit_price"]).sum()
avg_order_before = (sales["quantity"] * sales["unit_price"]).mean()
print(f"Total revenue: ${total_revenue_before:,.2f}")
print(f"Average order value: ${avg_order_before:,.2f}")

print("\n=== AFTER CLEANING (clean_sales) ===")
total_revenue_after = (clean_sales["quantity"] * clean_sales["unit_price"]).sum()
avg_order_after = (clean_sales["quantity"] * clean_sales["unit_price"]).mean()
print(f"Total revenue: ${total_revenue_after:,.2f}")
print(f"Average order value: ${avg_order_after:,.2f}")

print("\n=== REVENUE BY STATUS (clean_sales) ===")
revenue_by_status = (
    clean_sales.groupby("status")
    .apply(lambda g: (g["quantity"] * g["unit_price"]).sum(), include_groups=False)
    .rename("total_revenue")
    .reset_index()
)
display(revenue_by_status)

print("\n=== COMPARISON ===")
print(f"Total revenue difference: ${total_revenue_before - total_revenue_after:,.2f}")
print(f"Average order value difference: ${avg_order_before - avg_order_after:,.2f}")
print(f"Rows before: {len(sales)}")
print(f"Rows after: {len(clean_sales)}")

=== BEFORE CLEANING (original sales) ===
Total revenue: $42,830.94
Average order value: $140.43

=== AFTER CLEANING (clean_sales) ===
Total revenue: $34,373.19
Average order value: $146.27

=== REVENUE BY STATUS (clean_sales) ===


,status,total_revenue
0,COMPLETE,319.15
1,CANCELLED,437.83
2,Cancelled,5367.14
3,Cancelled,39.83
4,Complete,23774.72
5,Returned,2356.73
6,Returned,917.91
7,complete,309.87
8,completed,38.50
9,return,50.94



=== COMPARISON ===
Total revenue difference: $8,457.75
Average order value difference: $-5.84
Rows before: 305
Rows after: 235


## Data Quality Report

### Dataset Overview
- Original rows: (paste your number)
- Clean rows: (paste your number)
- Rows removed: (paste your number)

### Issues Found
- Missing customer IDs: (your number)
- Negative revenue: (your number)
- Zero quantity: (your number)
- Future dates: (your number)
- Duplicate order IDs: (your number)
- Invalid unit price: (your number)
- Null revenue: (your number)

### Revenue Comparison
- Total revenue before cleaning: (your number)
- Total revenue after cleaning: (your number)
- Average order value before: (your number)
- Average order value after: (your number)

### What Changed?
Removing problematic rows reduced total revenue and changed the average order value 
because outliers, negative values and invalid records were excluded.

### What Was Removed and Why?
- Rows with missing customer IDs (cannot identify the customer)
- Rows with negative or null revenue (invalid transactions)
- Rows with zero quantity (no product was sold)
- Rows with future dates (orders cannot exist in the future)
- Duplicate order IDs (same order counted twice)
- Invalid unit prices (zero or negative prices are not valid)


---

## Export cleaned dataset and report to the appropriate folder

In [18]:
Path('data/processed').mkdir(parents=True, exist_ok=True)



DATA QUALITY REPORT

Dataset Overview
Original rows: 305
Clean rows: 235
Rows removed: 70

Revenue Comparison
Total revenue before cleaning: $42,830.94

Total revenue after cleaning: $34,373.19

Average order value before: $140.43

Average order value after: $146.27

Things that changed
Removing problematic rows reduced total revenue and changed 
the average order value because outliers, negative values 
and invalid records were excluded.

Things that were removed
- Rows with missing customer IDs
- Rows with negative or null revenue
- Rows with zero quantity
- Rows with future dates
- Duplicate order IDs
- Invalid unit prices

---

### Reflection Questions.

Answer the reflection questions from the README below in markdown cells.

The biggest impact came from negative revenue and outlier 
values. These threw off the total revenue by $8,457.75 
which is a significant difference that could lead to 
bad business decisions if left unchecked.

Some assumptions I made during cleaning includes the fact that any row with a future date is invalid since orders 
cannot exist in the future, negative revenue is not a valid transaction, missing customer IDs make the order untraceable so they were removed, and duplicate order IDs mean the same order was counted more than once.

The risks that exist if the data validation step is skipped is that Skipping data validation means leadership is making 
decisions based on numbers that are just wrong.